src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -127,6 +127,26 @@ bool ShenandoahConcurrentGC::collect(GCCause::Cause cause) {
   // Complete marking under STW, and start evacuation
   vmop_entry_final_mark();
 
+  // If the GC was cancelled just before final mark (but after the preceding cancellation check),
+  // then the safepoint operation will do nothing and the concurrent mark will still be in progress.
+  // In this case it is safe (and necessary) to resume the degenerated cycle from the marking phase.
+  //
+  // On the other hand, if the GC is cancelled after final mark (but before this check), then the
+  // final mark safepoint operation will have finished the mark (setting concurrent mark in progress
+  // to false). In this case (final mark has completed), we need control to fall past the next
+  // cancellation check and resume the degenerated cycle from the evacuation phase.
+  if (heap->is_concurrent_mark_in_progress()) {
+    // If the concurrent mark is still in progress after the final mark safepoint, then the GC has
+    // been cancelled. The degenerated cycle must resume from the marking phase. Without this check,
+    // the non-generational mode may fall all the way to the end of this collect routine without
+    // having done anything (besides mark most of the heap). Without having collected anything, we
+    // can expect an 'out of cycle' degenerated GC which will again mark the entire heap. This is
+    // not optimal.
+    bool cancelled = check_cancellation_and_abort(ShenandoahDegenPoint::_degenerated_mark);
+    assert(cancelled, "GC must have been cancelled between concurrent and final mark");
+    return false;
```

Comment:
Not your change, but I notice that the semantics of the return value hasn't been explicitly documented anywhere in the class hierarchy. Might make sense to place a line of documentation, say in `ShenandoahGC` which first declares the virtual method, even something simple such as:

```
// Returns true if the collection was completed successfully, false otherwise.
```

Reply:
It also occurs to me that the pattern here is, roughly of the following form:

```
  do_some_uninterruptible_operation();
  check cancellation status & return if cancelled, else continue to next uninterruptible operation;
  ...
```

In that case, the checks should not be on the state of the progress of the collection, but on whether the cycle has been cancelled, and the asserts should then check the state of the collection; in other words, the dual of the structure of control in this method.

That said, I do not understand if there's a deeper rationale here in that sometimes we will want to skip a next uninterruptible phase of the collection regardless, because the state machine of the collection explicitly transitions us, under appropriate conditions, into a different phase than the otherwise normal flow, independent of whether a cancellation happened.

For example, what would be the difference between your current test:

```
  if (concurrent_mark_in_progress()) {
```

vs, say:

```
  if (!evacuation_in_progress()) {
```

vs : 

```
   if (check_cancellation()) {
     ... return false;
   }
   assert(!concurrent_mark_in_progress(), "Should have completed");
   assert(evacuation_in_progress(), "Next logical phase of collection");
```

Just wondering if a more uniform idiom might make for easier code patterns and easier, more uniform, reasoning about correctness.

Reply:
This case doesn't quite follow the `check_cancellation` pattern of the others because it is a little bit different. Here, the uninterruptible operation is the final mark safepoint. This safepoint will itself check if the mark has been cancelled. If final mark detects the cancellation, it will do _nothing_, the concurrent mark will still be in progress and we _must_ resume the degenerated cycle from the marking phase. However, if the final mark safepoint does _not_ detect a cancellation, it will initialize the evacuation phase. If the code simply `checks_cancellation` after final mark it cannot tell if the cancellation request was received just _before_ final mark or just _after_ final mark. These two conditions require different resumption points for the degenerated cycle, so we must check the state of the collection cycle here.

Reply:
Thanks for the explanation. I do wonder if there's a one-to-one bijection between the interrupt/cancellation detection point and the resumption point for the degeerated cycle. If that is true, then the resumption point may be identified with the state of the concurrent GC state machine (the degeneration work is in a coupled state machine in the sense of David Harel) and might avoid this more subtle and perhaps a little bit more fragile programming idiom. But we can think about this in the fullness of time.

Reply:
Yes, it would be great to clean up the state hand-off from concurrent to degenerated cycles. The generational mode also introduced a new `_degenerated_roots` state to avoid double-resetting of the write card table.